robustness edits and tests

tassiajones opened the pull request:
edited current R files to improve robustness and documentation. also created new test files for each file and have run and all tests passed.

tassiajones committed:
robustness edits and tests

edited current R files to improve robustness and documentation. also created new test files for each file and have run and all tests passed.

tassiajones committed:
robustness and documentation edits

updated existing scripts for robustness and documentation and ran new test scripts with no errors

tassiajones committed:
added result_id to example data

CYShapland commented on R/append_values.R:
```diff
@@ -5,7 +5,7 @@
 #' @param dat A bias dataset in long format (must include j, d, t, domain, result_id)
 #' @param values A data frame of prior distributions (e.g., from dat_bias_values)
-#' @param common Should priors apply across all domains (TRUE), or vary by domain?
+#' @param common Should a single set of distributions be used across all domains (default is TRUE)?
```
should very specific and "prior" before "distribution"

tassiajones replied:
im not 100% sure i understand what you mean here, should i change to something like this '#' @param common Should a single set of prior distributions be used across all domains (default is TRUE)? Set to FALSE to allow domain-specific priors.
'

CYShapland replied:
Yes exactly right.

CYShapland commented on R/append_values.R:
```diff
@@ -75,7 +75,7 @@ tri_append_bias <- function(dat, values, common = TRUE) {
 #' @param dat A long-format indirectness dataset
 #' @param values A dataframe of indirectness priors (e.g., dat_ind_values)
-#' @param common Should priors apply across all domains (TRUE), or vary by domain?
+#' @param common Should a single set of distributions be used across all domains (default is TRUE)?
```
Same as comment above

CYShapland reviewed (changes requested):
I've left some comments in the code and make sure you still keep Luke's annotations and explanations if possible

tassiajones committed:
added beta carotene data and made chin yang updates

have added the beta carotene data to the data in triangulate and have run it with the updated files and seems to give same results as the paper. also have made updates from chin yang comments.

tassiajones committed:
updated documentation

mcguinlu commented on R/analysis.R:
```diff
@@ -1,63 +1,96 @@
+    )
 
   dat_final <- dat_bias %>%
     dplyr::distinct(result_id, study, yi, vi) %>%
```
It would be useful if `tri_prep_data()` also returned the `type` variable, to allow for down-stream stratification by study type (if needed)

mcguinlu commented on R/analysis.R:
```diff
@@ -1,63 +1,96 @@
+      ) / (scale_factor^2)
     ) %>%
-    return()
+    # Replace Inf or NaN with NA and warn
```
Did you find a case where a divide-by-zero error occurred? IIRC, `propimn` and `propemn` should never be zero (and so `scale_factor` should never be zero) because they are the exponentiated sum of the bias and variance across domains, divided by two:

https://github.com/mcguinlu/triangulate/blob/b517c17112f5620cb2a5876b0698274a442a6e43/R/analysis.R#L68-L71

So even if the bias across all domains is 0, then the sum of these is also 0, which gives a `propimn` of

```
exp(0+0/2) =1
```

mcguinlu commented on tests/testthat/test-analysis.R:
```diff
@@ -0,0 +1,17 @@
+  result <- tri_prep_data(dat_bias_prep, dat_ind_prep)
+
+  expect_true("yi_adj" %in% colnames(result))
+  expect_true("vi_adj" %in% colnames(result))
```
Add end-to-end test here - see overall review comment.

mcguinlu reviewed (dismissed):
Hi @tassiajones - thanks for the excellent PR. You've done a great job at cleaning up the code and making it a bit more user-friendly - I particularly like the informative error handling you've built in. I've made a couple of comments for your review.

One overall suggestion is that you add a further test check the output of the full data cleaning/analysis pipeline against a 'known' result - essentially a test to confirm that given a set of results and priors, the package reliably generates the correct final dataset. This will help us have confidence as we add more features. Maybe @CYShapland could provide a final, cleaned dataset she has confidence in to use in this test?

CYShapland commented on R/dat_check.R:
```diff
@@ -1,23 +1,30 @@
-  cols <- colnames(dat)
+#' @return Throws an error if required columns are missing. Otherwise returns TRUE (invisibly).
+#' @export
+tri_dat_check <- function(dat, mode = c("minimal", "full")) {
```
So default is not "minimal"?

CYShapland commented:
@mcguinlu we will pull some of robviz function and change it slightly for triangulation exercise, adding different method of meta-analysis, including very high risk of bias and creating figures like the ones in our manuscript.

tassiajones commented:
Hi Luke,

Thanks very much for your feedback that's super helpful!

Regarding the suggestion about adding a data set for the end-to-end test, I have actually added the Beta Carotene dataset that Chin Yang uses in her paper. I am currently in the process of adding the full test file and tutorial based on this dataset.

Hope this helps.

Thanks,

Tassia.
________________________________
From: Luke McGuinness ***@***.***>
Sent: 10 July 2025 13:32
To: mcguinlu/triangulate ***@***.***>
Cc: Tassia Jones ***@***.***>; Mention ***@***.***>
Subject: Re: [mcguinlu/triangulate] robustness edits and tests (PR #7)


@mcguinlu requested changes on this pull request.

Hi @tassiajones<https://github.com/tassiajones> - thanks for the excellent PR. You've done a great job at cleaning up the code and making it a bit more user-friendly - I particularly like the informative error handling you've built in. I've made a couple of comments for your review.

One overall suggestion is that you add a further test check the output of the full data cleaning/analysis pipeline against a 'known' result - essentially a test to confirm that given a set of results and priors, the package reliably generates the correct final dataset. This will help us have confidence as we add more features. Maybe @CYShapland<https://github.com/CYShapland> could provide a final, cleaned dataset she has confidence in to use in this test?

________________________________

In R/analysis.R<https://github.com/mcguinlu/triangulate/pull/7#discussion_r2197547773>:

>
   e_prop <- dat_ind %>%
     dplyr::group_by(result_id) %>%
-    dplyr::summarise(sumlogmn = sum(ind_m_prop, na.rm = T),
-              sumlogvr = sum(ind_v_prop, na.rm = T),
-              propemn = exp(sumlogmn+sumlogvr/2),
-              propevar = (exp(2*sumlogmn+sumlogvr)*(exp(sumlogvr)-1))) %>%
-    dplyr::select(result_id, dplyr::starts_with("prop"))
+    dplyr::summarise(
+      sumlogmn = sum(ind_m_prop, na.rm = TRUE),
+      sumlogvr = sum(ind_v_prop, na.rm = TRUE),
+      propemn = exp(sumlogmn + sumlogvr / 2),
+      propevar = exp(2 * sumlogmn + sumlogvr) * (exp(sumlogvr) - 1),
+      .groups = "drop"
+    )

   dat_final <- dat_bias %>%
     dplyr::distinct(result_id, study, yi, vi) %>%


It would be useful if tri_prep_data() also returned the type variable, to allow for down-stream stratification by study type (if needed)

________________________________

In R/analysis.R<https://github.com/mcguinlu/triangulate/pull/7#discussion_r2197590296>:

>      ) %>%
-    return()
+    # Replace Inf or NaN with NA and warn


Did you find a case where a divide-by-zero error occurred? IIRC, propimn and propemn should never be zero (and so scale_factor should never be zero) because they are the exponentiated sum of the bias and variance across domains, divided by two:

https://github.com/mcguinlu/triangulate/blob/b517c17112f5620cb2a5876b0698274a442a6e43/R/analysis.R#L68-L71

So even if the bias across all domains is 0, then the sum of these is also 0, which gives a propimn of

exp(0+0/2) =1


________________________________

In tests/testthat/test-analysis.R<https://github.com/mcguinlu/triangulate/pull/7#discussion_r2197604302>:

> @@ -0,0 +1,17 @@
+test_that("tri_prep_data() and tri_calculate_adjusted_estimates() run correctly", {
+  dat_bias_prep <- triangulate::dat_bias %>%
+    tri_to_long() %>%
+    tri_absolute_direction() %>%
+    tri_append_bias(triangulate::dat_bias_values)
+
+  dat_ind_prep <- triangulate::dat_ind %>%
+    tri_to_long() %>%
+    tri_absolute_direction() %>%
+    tri_append_indirect(triangulate::dat_ind_values)
+
+  result <- tri_prep_data(dat_bias_prep, dat_ind_prep)
+
+  expect_true("yi_adj" %in% colnames(result))
+  expect_true("vi_adj" %in% colnames(result))


Add end-to-end test here - see overall review comment.

—
Reply to this email directly, view it on GitHub<https://github.com/mcguinlu/triangulate/pull/7#pullrequestreview-3005404186>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A5XKHTZDYFXVBXAKLBP7SHL3HZMPLAVCNFSM6AAAAACA4QITZKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMBVGQYDIMJYGY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

tassiajones committed:
added in an example using the beta carotene dataset

have also added a test file for the new plot direction bias function which uses robvis structure from beta carotene github code. have put the example in inst as it wasnt a full vignette but demo is actually outdated package practice. let me know if that needs to change/

tassiajones committed:
moved the beta carotene example to vignettes/articles

note that it uses library(triangulate) but when i run it i use the devtools::load_all() function to ensure i am using my current branch of the package and not the original package functions

tassiajones committed:
added imported packages to description

tassiajones committed:
fixing pkgdown error

tassiajones committed:
added sensitivity analysis vignettes

added how to analyse how sensitive results are to prior beliefs and plot them, also did an interactive shiny plot so users can see how their priors affect results easily and did a vignette for this. also added me and chin yang as authors

CYShapland commented on R/data.R:
```diff
@@ -26,12 +26,4 @@
 #' Cleaned dataset of log risk ratios for CHD, merged with RoB annotations.
 
-#' @format A data frame with 13 rows and ~25 variables
-#' @source Shapland et al. (2024)
```
Is it possible to add a link of this paper somewhere? Maybe at the beginning of the example?

CYShapland commented on inst/example/CHD/01_CHD_Prep_Data.R:
I would remove this, as this is not part of the triangulate process, may confuse people, this is more transforming different exposures to harmonise them, not to do with triangulate. Just save the cleaned data "Res_MetaAll_beta_carotene_AllCHD_v2" within the example data folder

CYShapland commented on R/plot_bias_direction.R:
```diff
@@ -0,0 +1,441 @@
+    #-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-
+    # Make forest plot
+
+    res_all <- rma(yi, vi, data=dat)
```
could you also give people an option for this function to choose what "methods" for the meta-analysis i.e. the rma()

CYShapland commented on R/plot_bias_direction.R:
```diff
@@ -0,0 +1,441 @@
+
+    par(fig=c(0.6,1,0,1), new=TRUE)
+
+    res_adj <- rma(yi_adj, vi_adj, data=dat_adj)
```
Same here, could you add a methods option

CYShapland commented on R/plot_bias_direction.R:
```diff
@@ -0,0 +1,441 @@
+        vi_adj,
+        subset = (type == x),
+        data = dat_adj,
+        #method = "DL"  ### CHANGE to have "REML" the default rma
```
Same here again about having options for methods

CYShapland commented on R/plot_bias_direction.R:
Could you test this on luke's example data, I'm a little concerned this code I wrote a while ago may not work on all datasets

CYShapland commented on vignettes/articles/beta-carotene-chd-example.Rmd:
```diff
@@ -0,0 +1,202 @@
+
+This tutorial demonstrates how to use the `triangulate` package to perform a bias-adjusted meta-analysis of the association between beta-carotene and coronary heart disease (CHD), using estimates from diverse study designs (RCTs, observational, and MR).
+
+## 1. Create CHD Dataset
```
I removed this section, as this is not something triangulate can do

CYShapland committed:
Update beta-carotene-chd-example.Rmd

CYShapland commented on vignettes/articles/beta-carotene-chd-example.Rmd:
````diff
@@ -0,0 +1,204 @@
+```
+
+## 4. Add Indirectness (set to missing)
+
````
Could you add a note to say that indirectness is not necessary always missing, only for this particular example. And the title of this section needs to add adjust effect estimate as "tri_append_indirect()" also adjusts the effect estimate based on the indirectness results

CYShapland replied:
Nice, I would change the last section of the sentence to "indirectness judgments can and should be specified if available." And also the last section of the last sentence to "therefore the pipeline adjusts the effect estimate by both bias and indirectness"

CYShapland commented on vignettes/articles/beta-carotene-chd-example.Rmd:
````diff
@@ -0,0 +1,204 @@
+## 5. Apply Bias Priors & Estimate Adjusted Effects
+
+```{r bias-correction}
+custom_bias_priors <- triangulate::dat_bias_values %>%
````
This needs a bit more explanation. A sub-section to say this is the customised prior added to the default prior

CYShapland commented on vignettes/articles/beta-carotene-chd-example.Rmd:
```diff
@@ -0,0 +1,204 @@
+  tri_absolute_direction() %>%
+  tri_append_bias(custom_bias_priors)
+
+datCHD_final <- tri_prep_data(dat_CHDbias_prep, dat_CHDind)
```
I think this line belongs to a new section, as this is the final dataset that combines bias and indirectness adjustments

CYShapland replied:
I have changed this directly

CYShapland commented on vignettes/articles/beta-carotene-chd-example.Rmd:
````diff
@@ -0,0 +1,204 @@
+```{r plot, fig.width=10, fig.height=8}
+
+### Plot adjusted with robviz
+dat_CHDbias_final <- dat_CHDbias_rob %>%
````
What is "dat_CHDbias_final" for? looks like the same code as below?

CYShapland commented on vignettes/articles/beta-carotene-chd-example.Rmd:
```diff
@@ -0,0 +1,204 @@
+dat_CHDbias_final <- dat_CHDbias_rob %>%
+  right_join(datCHD_final[c("result_id", "yi_adj", "vi_adj")], by = "result_id")
+
+dat_plot <- left_join(dat_CHDbias, datCHD_final[, c("result_id", "yi_adj", "vi_adj")])
```
Could you create a function that automatically does  "left_join(dat_CHDbias, datCHD_final[, c("result_id", "yi_adj", "vi_adj")])" into the "tri_plot_bias_direction()"?

CYShapland commented on vignettes/articles/sensitivity-analysis.Rmd:
```diff
@@ -0,0 +1,167 @@
+
+# this is where we combine the bias/indirectness priors with the study results and adjusts yi and vi to get new columns yi_adj and vi_adj for the adjusted estimate and varainces, respectively
+
+base_prepped <- tri_prep_data(bias_priors, indirectness_priors)
```
Ooo, where is these two commands?

tassiajones replied:
they were manually defined in the code section above

CYShapland reviewed (changes requested):
Looking great, I've added some comments and wish lists. We can discuss further on Monday :)

tassiajones committed:
adjusting the plotting function and trying to make it more reproducible

edited the tri_plot_bias_dierction() function as it was cutting off some of the plots. also tried to make it more reproducible with a variety of number of studies by dynamically scaling the vertical buffer based on the number of studies and subgroups. not sure how much difference it makes but i have added a vignette which experiments with example datasets. i will change the format of everything as theres now 2 plotting functions but let me know what you think.

tassiajones committed:
fixed pkgdown

CYShapland commented on R/plot_bias_direction.R:
```diff
@@ -6,12 +6,23 @@
 #' @param title Optional plot title.
 #' @param grouping Column by which to stratify subgroups (default = `"type"`).
+#' @param dat_adj A data frame with adjusted estimates and variances, if different from the main input.
+#' @param rma_method Meta-analysis method passed to metafor::rma (e.g., "REML", "FE").
```
could you just add here we use metafor's default which is "random-effects"

CYShapland commented on vignettes/articles/beta-carotene-chd-example.Rmd:
```diff
@@ -25,95 +25,26 @@ library(magrittr)
+## 1. Load CHD Dataset
 
-We define estimates from different study designs and harmonise them to a common dose-response scale.
+The Metadta_all_CHD dataset compiles results from observational, biomarker, randomised controlled trial (RCT), and Mendelian randomisation (MR) studies examining the association between beta-carotene exposure and coronary heart disease (CHD). It includes standardised effect estimates (log risk ratios) and standard errors, allowing for dose-response meta-analysis across different study designs.
```
biomarker? is it the data from my paper? we didn't have a biomarker study

tassiajones replied:
sorry! was doing some more reading for the presentation and think i had put it in here by accident - will change now.

CYShapland committed:
Update beta-carotene-chd-example.Rmd

CYShapland commented on vignettes/articles/reproducible-plotting-function.Rmd:
Should this be part of the test files rather than in the article?

tassiajones replied:
yes it should - i just wanted to upload it in an easy format for you to have a look at before i did it as a test file. do you think it looks okay?

CYShapland replied:
Oh I see, yes they look great, just one thing, the "O" section is missing in all the plots, is that because you didn't define it? (sorry if I missed this). There is also a wide gap between the headers and the first subgroup subtitle, is it possible to reduce this? for example the [tri_plot_20_studies.png], the gap between "Study" and "RCT" is quite wide.

tassiajones replied:
yes will change the spacing now. and yes sorry i didnt define it in the example data i will edit it so it does.

tassiajones committed:
updated some documentation and notes and the plotting function should work for any amount of studies now

tassiajones committed:
fixing pkgdown
